go-mode.el

```diff
@@ -1995,15 +2001,13 @@ arguments can be set as a list via ‘gofmt-args’."
       (kill-buffer errbuf))))
 
 ;;;###autoload
-(defun gofmt-before-save ()
-  "Add this to .emacs to run gofmt on the current buffer when saving:
-\(add-hook 'before-save-hook 'gofmt-before-save).
-
-Note that this will cause ‘go-mode’ to get loaded the first time
-you save any file, kind of defeating the point of autoloading."
-
-  (interactive)
-  (when (eq major-mode 'go-mode) (gofmt)))
+(define-minor-mode gofmt-mode
```

Comment:
The minor mode approach is what the prettier mode does as well. It was where I got the idea.

https://github.com/prettier/prettier-emacs/blob/e9b73e81d3e1642aec682195f127a42dfb0b5774/prettier-js.el#L203-L210